Optimized PointInstance importing

vickycl opened the pull request:
## Purpose of this PR

**Ticket/Jira #:** N/A

<!-- Description of feature/change. Links to screenshots, design docs, user docs, etc. Remember reviewers may be outside your team, and not know your feature/area that should be explained more. -->

PR #331 introduced lots of Transform.Find calls when importing a PointInstancer, which in initial testing didn't appear to introduce a regression but has since proved to be problematic. Discussing with @John-Whigham, we discovered this is because the order of the transform hierarchy means one behaviour is best case (finding each child immediately, deleting it, then recreating on the end) while the other is worst case (not finding the child so searching to the end, with each iteration larger than the last). In profiling this divided loop proves to be much faster, when we destroy all the old objects first before re-instantiating, in all cases.

## Testing

**Functional Testing status:**

Tested on Pixar's PointInstancedMedCity file.

**Performance Testing status:**

<!-- Could this PR affect performance? If so, what has been done to measure time taken / memory used etc? Can include new or existing performance tests. Also see [Ensuring Performance by Default](https://confluence.unity3d.com/display/DEV/Ensuring+Performance+by+Default). -->

Before: 
![image](https://user-images.githubusercontent.com/45685026/216310575-3b5f4d29-c08b-42ea-9185-fa49a965d23e.png)


After:
![image](https://user-images.githubusercontent.com/45685026/216310766-3a3b3461-7676-4062-a37a-454d0fa9cd31.png)

Tested that initial import, re-import and refresh from USD no longer take a ridiculous time (>1minute) both in a cold and warm editor.


## Overall Product Risks
<!-- See Testing Status, Complexity and Halo Effect for your PR](https://confluence.unity3d.com/display/DEV/Testing+Status%2C+Complexity+and+Halo+Effect+for+your+PR). -->

**Complexity:** low
<!-- (Minimal / Low / Medium / High) -->

**Halo Effect:** low
<!-- (Minimal / Low / Medium / High) -->

## Additional information

**Note to reviewers:**

<!-- Info per person for what to focus on, or historical info to understand who have previously reviewed and coverage. Help them get context. -->

**Reminder:**
<!-- Things to remember to do before finalizing this PR. -->
- [ ] Add entry in CHANGELOG.md _(if applicable)_

vickycl committed:
Optimized PointInstance importing

Destroy all the old objects first before re-instantiating to have make finding objects in the transform hierarchy faster.

Commit:
Apply formatting changes

michaeljblain commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/InstanceImporter.cs:
```diff
@@ -98,6 +98,8 @@ public static void BuildSceneInstances(PrimMap primMap, SceneImportOptions optio
             */
 
+            Tuple<Matrix4x4, string, GameObject>[] instancesToCreate = new Tuple<Matrix4x4, string, GameObject>[sample.protoIndices.Length];
+            int i = 0;
```
Minor: we should call this something more descriptive e.g. instanceCount

michaeljblain reviewed (changes requested):
Alongside my minor suggestion I think it would be worth adding a comment for each loop describing what it is doing.

vickycl committed:
Renamed variable and added comments.

vickycl committed:
Remove unused inactiveIds.

vickycl committed:
Remove debug warning spam.

When a prototype isn't in the prim map, a warning is written to the console for every instance which is unnecessary and also very harmful for performance. It is already logged once for each prototype which is enough.

lee-aandrew commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/InstanceImporter.cs:
```diff
@@ -68,14 +69,13 @@ public static void BuildSceneInstances(PrimMap primMap, SceneImportOptions optio
             Matrix4x4[] transforms = sample.ComputeInstanceMatrices(scene, pointInstancerPath);
-            int i = 0;
 
             foreach (var protoRoot in sample.prototypes.targetPaths)
```
Does "proto" here mean prototype?
could we maybe change this to "prototype" and for below "prototypeRoot" if so?

vickycl replied:
Sure :)

lee-aandrew commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/InstanceImporter.cs:
```diff
@@ -117,36 +123,45 @@ public static void BuildSceneInstances(PrimMap primMap, SceneImportOptions optio
                 }
 
-                var goInstance = GameObject.Instantiate(goMaster, root.transform);
+                instancesToCreate[instanceCount] = new Tuple<Matrix4x4, string, GameObject>(xf, instanceName, goMaster);
```
can we have the tuple here named?
so that we dont have to use Item1, Item2, Item3 when referencing them?

vickycl replied:
TIL about ValueTuples that can actually have named fields 😁 Done

vickycl committed:
Updates from review.

vickycl committed:
Merge branch 'dev' into fix-point-instancer-perf-bug
